pkg/tokenvalidation/validationconfig.go

```diff
@@ -0,0 +1,165 @@
+	*completedConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *TokenValidationServerConfig) Complete() completedConfig {
+	cfg := completedConfig{
+		GenericConfig:             c.GenericConfig.Complete(),
+		ExtraConfig:               &c.ExtraConfig,
+		kubeAPIServerClientConfig: c.GenericConfig.ClientConfig,
+	}
+
+	return cfg
+}
+
+// New returns a new instance of TokenValidationServer from the given config.
+func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*TokenValidationServer, error) {
+	genericServer, err := c.GenericConfig.New("oauth.openshift.io-tokenvalidation", delegationTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &TokenValidationServer{
+		GenericAPIServer:             genericServer,
+		AccessTokenInactivityTimeout: c.ExtraConfig.AccessTokenInactivityTimeout,
+		kubeAPIServerClientConfig:    c.kubeAPIServerClientConfig,
+	}
+
+	s.installTokenValidation()
+
+	return s, nil
+}
+
+func (s *TokenValidationServer) installTokenValidation() error {
+	// build clients
+	userClient, err := userclients.NewForConfig(s.GenericAPIServer.LoopbackClientConfig)
+	if err != nil {
+		return err
+	}
+
+	oauthClient, err := oauthclients.NewForConfig(s.GenericAPIServer.LoopbackClientConfig)
+	if err != nil {
+		return err
+	}
+
+	corev1Client, err := corev1.NewForConfig(s.kubeAPIServerClientConfig)
+	if err != nil {
+		return err
+	}
+	bootstrapUserDataGetter := bootstrap.NewBootstrapUserDataGetter(corev1Client, corev1Client)
+
+	// create informer for the users to be used in user <-> groups mapping
+	userInformer := userinformer.NewSharedInformerFactory(userClient, defaultInformerResyncPeriod)
+	if err := userInformer.User().V1().Groups().Informer().AddIndexers(cache.Indexers{
+		usercache.ByUserIndexName: usercache.ByUserIndexKeys,
+	}); err != nil {
+		return err
+	}
+
+	groupMapper := usercache.NewGroupCache(userInformer.User().V1().Groups())
+	oauthInformer := oauthinformer.NewSharedInformerFactory(oauthClient, defaultInformerResyncPeriod)
+
+	timeoutValidator := tokenvalidators.NewTimeoutValidator(
+		oauthClient.OauthV1().OAuthAccessTokens(),
+		oauthInformer.Oauth().V1().OAuthClients().Lister(),
+		s.AccessTokenInactivityTimeout,
+		minimumInactivityTimeoutSeconds)
+
+	s.GenericAPIServer.AddPostStartHookOrDie("openshift.io-StartUserInformer", func(ctx genericapiserver.PostStartHookContext) error {
+		go userInformer.Start(ctx.StopCh)
+		return nil
+	})
+	s.GenericAPIServer.AddPostStartHookOrDie("openshift.io-StartOAuthInformer", func(ctx genericapiserver.PostStartHookContext) error {
+		go oauthInformer.Start(ctx.StopCh)
+		return nil
+	})
+	s.GenericAPIServer.AddPostStartHookOrDie("openshift.io-StartTokenTimeoutUpdater", func(ctx genericapiserver.PostStartHookContext) error {
+		go timeoutValidator.Run(ctx.StopCh)
+		return nil
+	})
+
+	validationHandler := NewTokenValidationHandler(
```

Comment:
I suggest adding a strategy like we did for the TokenReviews upstream.  Doing it that way avoids adding all this and simply adds a new resource to where you're building your existing one.

You could have

`/apis/oauth.openshift.io/v1/oauthtokenreviews/`

Reply:
See https://github.com/kubernetes/kubernetes/blob/07d88617e55b64c243956c7b2032430beb03b159/pkg/registry/authentication/tokenreview/storage.go for example.

Reply:
@deads2k  I haven't looked much into generating the openapi models needed in order to be able to directly serve the actual k8s.TokenReview object needed by the rest endpoint, but the other way around (mirroring the TokenReview object in the openshift/api) proved unfruitful.

I used the type from my https://github.com/stlaz/api/commit/584f090a52af86aac4b61067ab7730c446f031c8, however, after wiring it similarly to what upstream does (basically just replacing the type in the `REST.New()` method), the generic handlers eventually won't pass the object to `REST.Create()` because the `gvk` of the object sent by kube does not match (the groups differ) - [kube-apiserver crete handler gvk check](https://github.com/kubernetes/apiserver/blob/0e46f0ea2bdc958827712205cf9a88140a28409d/pkg/endpoints/handlers/create.go#L125).

That leads me to believe that some kind of a prelude handler might always be necessary. I could hack around the generic API machinery handlers by keeping the below code but instead of `validationHandler` have a wrapper that would call the upstream's `tokenreview.REST.Create()` with the object from the incoming request. If that's what you'd prefer.

Or maybe you had something simpler in mind, my solutions may oftentimes be too complicated simply because I don't know what to look for.